### DISCUSSION TITLE
How to catch errors during pre-rendering in Next.js?

### DISCUSSION THREAD

**Discussion by hanshank** [vercel/next.js#39264] [Help]
Hi.

My team and I are looking into catching runtime errors during server-side rendering/pre-rendering in Next.js. I'm aware that the [ErrorBoundary](https://reactjs.org/docs/error-boundaries.html) strategy that the React docs suggests for client-side errors, will not work for this.

**Triggering a sample error:**

```js
// imagine this inside of pages/some-page.js
function Page() {
  // this will show "ReferenceError: window is not defined" in the terminal in production mode, or in an overlay in dev mode
  console.log(window);

  return <div>Some Page</div>
}
 ```

**My question is:** How can I catch errors such as the above, server-side (during pre-rendering) and do some custom logic, such as logging or showing a specific error page?

**Comment by hanshank** [comment]
Any thoughts on this?

**Comment by gaston-flores** [comment]
For client/server-side errors [you can override the `_error.js` component](https://nextjs.org/docs/advanced-features/custom-error-page#more-advanced-error-page-customizing) to show a custom message or do something with the error in `getInitialProps`. 

If you are interested in a managed solution, [Sentry has a Next.js integration](https://docs.sentry.io/platforms/javascript/guides/nextjs/) that works really well.

**Comment by hanshank** [comment]
Thanks @gaston-flores ! Sentry might be a helpful tool, but my organization does have other services for logging and error tracking that we are hoping to use instead.

Do you know what the hierarchy/execution for 500.js and 404.js and the _error.js pages are? My understanding is that for 500 and 404 errors, the default templates will be used unless actual templates are specified in the code base. When will the _error.js page come into play?

**Reply by gaston-flores** [reply]
Both pages override the `_error.js` page, but `_error.js` overrides the default templates. So it goes:

1. Custom `500.js` and `404.js`
2. Custom `_error.js`
3. Default error pages

If you are only interested in 500 errors you can create `404.js` and `_error.js`. Next will only use your `_error.js` for 500 errors in that case.

**Comment by olivierr91** [comment]
@gaston-flores The link you provide for overriding _error.js works only for the old Pages router, how can we do this using the App router? Having the ability to log errors server-side when a page fails to render server side is important for us.
